adr/graph-feature-flags.md

```diff
@@ -0,0 +1,212 @@
+
+### Feature Flags must not break composition checks
+
+At any time, all existing feature flags must be compatible with all existing subgraphs,
+based on the label matching rules of the subgraphs and feature flags.
+
+### Feature Flag names must be unique across flags and groups
+
+Within the same namespace, feature flags and group feature flags must have unique names.
+It's not allowed to have a group feature flag with the same name as a feature flag and vice versa.
+It's not allowed to have a group feature flag with the same name as another group feature flag.
+It's not allowed to have a feature flag with the same name as another feature flag.
+
+### Feature Flags are namespaced
+
+Feature Flags are namespaced.
+
+### The Router Engine Config is aware of all Feature Flags
+
+As we explicitly define through label matching which feature flags are active for a federated graph,
+we will statically embed all possible configurations into the Router Engine Config.
+The engine config will be an array with a key:value pair for the feature flag, and another key:value pair for the config.
+The config will always contain a default configuration that does not contain any feature flag key.
+
+### Router Handling of Requests with Feature Flags
+
+If a request contains a feature flag, the Router will check if the feature flag exists in the Router Engine Config.
+If it does, the Router will execute the request with the feature flag configuration.
+If it does not, the Router will return a 400 Bad Request error.
+
+### Analytics and Metrics need to take into consideration Feature Flags
+
+Both Metrics and Tracing need to be aware of feature flags.
+Each entry needs to contain the feature flag that was used for the request so that one can filter by feature flag.
+It's a requirement to be able to understand metrics like error rates, latency, and throughput per feature flag.
+
+### The Studio should show all Feature Flags
+
+The Studio should show all feature flags that are available for a federated graph.
+
+### The Playground needs to be aware of Feature Flags
+
+On the Playground, it should be possible to see all available feature flags for a federated graph.
+The Playground should allow selecting a feature flag from a dropdown.
+The Playground should pre-select the default feature flag, which means that no feature flag is active.
+Based on the selected feature flag, the Playground should set the appropriate header or cookie.
+
+### Feature Flags must work for all types of requests
+
+- Subscriptions over WebSocket (header, cookie, extensions)
+- Queries and Mutations over HTTP (header, cookie, extensions)
+- Subscriptions over SSE (header, cookie, extensions)
+
+### Introspection needs to be aware of Feature Flags
+
+Depending on the selected feature flag, the introspection query should return the schema that is associated with the feature flag.
+
+### No security through obscurity
```

Comment:
will there be a setting to disable feature flag functionality i.e. only enable it on the dedicated federated graph and not on others? Meaning if enabled the flag is respected on the request but if disabled it returns a 400?

Reply:
Yes, this is a good idea. We can add it to the Router Config.

Reply:
Great!